### PR TITLE
Add Blue Pearl Visual Verification Suite

### PR DESCRIPTION
**Pull request by SimonMatthews-BP**
Add support for Blue Pearl. Latest changes were fixes to code style issues.  

- [ X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X ] Ensure that the pull request title represents the desired changelog entry
- [ X] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ X] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/edu/hm/hafner/analysis/parser/BluePearlParser.java
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/BluePearlParser.java:58]: This method is only intended to be overwritten to **speed up** parsing of large files. It should filter uninteresting lines that slow down parsing when the general regexp is applied. So it should only contain string comparisons. 

If you want to use this line for filtering of lines that should not match, then please update your regexp or the `createIssue` method.

File: src/main/java/edu/hm/hafner/analysis/registry/BluePearlDescriptor.java
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/registry/BluePearlDescriptor.java:12]: Since you capitalized the parser `BluePearl` it might make sense to use as ID `blue-pearl`. But I leave that up to you...

File: src/main/java/edu/hm/hafner/analysis/registry/ParserRegistry.java
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/registry/ParserRegistry.java:107]: Can you insert alphabetically?

File: src/main/java/edu/hm/hafner/analysis/parser/BluePearlParser.java
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/BluePearlParser.java:47]: Maybe it is sufficient to look at all files that start not with `--`?

**Review by uhafner — CHANGES_REQUESTED** [review]

**Comment by SimonMatthews-BP** [comment]
On Sun, Jun 12, 2022 at 12:50 PM Ullrich Hafner ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In src/main/java/edu/hm/hafner/analysis/parser/BluePearlParser.java
> <https://github.com/jenkinsci/analysis-model/pull/803#discussion_r895225180>
> :
>
> > +            Pattern fileAndLineNo = Pattern.compile("[a-zA-Z-0-9._]+\\(\\d+\\)");
> +            Matcher m = fileAndLineNo.matcher(line);
> +            if (m.find()) {
> +                return line.startsWith("W-") || line.startsWith("E-") || line.startsWith("F-");
> +            }
> +            else {
> +                return false;
> +            }
>
> This method is only intended to be overwritten to *speed up* parsing of
> large files. It should filter uninteresting lines that slow down parsing
> when the general regexp is applied. So it should only contain string
> comparisons.
>
I'll re-work this.


> ------------------------------
>
> In src/main/java/edu/hm/hafner/analysis/registry/BluePearlDescriptor.java
> <https://github.com/jenkinsci/analysis-model/pull/803#discussion_r895225393>
> :
>
> > @@ -0,0 +1,23 @@
> +package edu.hm.hafner.analysis.registry;
> +
> +import edu.hm.hafner.analysis.IssueParser;
> +import edu.hm.hafner.analysis.parser.BluePearlParser;
> +
> +/**
> + * A descriptor for the Blue Pearl Visual Verification Suite.
> + *
> + * @author Simon Matthews
> + */
> +class BluePearlDescriptor extends ParserDescriptor {
> +    private static final String ID = "bluepearl";
>
> Since you capitalized the parser BluePearl it might make sense to use as
> ID blue-pearl. But I leave that up to you...
>
I would prefer to leave it as "bluepearl".

> ------------------------------
>
> In src/main/java/edu/hm/hafner/analysis/registry/ParserRegistry.java
> <https://github.com/jenkinsci/analysis-model/pull/803#discussion_r895225441>
> :
>
> > @@ -104,6 +104,7 @@ public class ParserRegistry {
>              new KtLintDescriptor(),
>              new MavenConsoleDescriptor(),
>              new MentorGraphicsDescriptor(),
> +            new BluePearlDescriptor(),
>
> Can you insert alphabetically?
>
Will do.


> ------------------------------
>
> In src/main/java/edu/hm/hafner/analysis/parser/BluePearlParser.java
> <https://github.com/jenkinsci/analysis-model/pull/803#discussion_r895225734>
> :
>
> > +    @Override
> +    protected Optional<Issue> createIssue(final Matcher matcher, final LookaheadStream lookahead,
> +                                          final IssueBuilder builder) {
> +
> +        builder.guessSeverity(matcher.group("severity"));
> +        builder.setMessage(matcher.group("messageString"));
> +        builder.setFileName(matcher.group("filename"));
> +        builder.setLineStart(matcher.group("lineNo"));
> +        builder.setCategory(matcher.group("type"));
> +
> +        return builder.buildOptional();
> +    }
> +
> +    @Override
> +    protected boolean isLineInteresting(final String line) {
> +        if (line.startsWith("--BPS-0730")) {
>
> Maybe it is sufficient to look at all files that start not with --?
>
Good point. I'll change this.

File: src/main/java/edu/hm/hafner/analysis/parser/BluePearlParser.java
Comment (SimonMatthews-BP) [reply src/main/java/edu/hm/hafner/analysis/parser/BluePearlParser.java]: I have simplified the code and eliminated regexes from this function.

**Comment by SimonMatthews-BP** [comment]
Please hold off the pull request. I think I need to update the way the warning severities are added: I don't think the "guessSeverity" is doing a good job for this parser.

File: src/main/java/edu/hm/hafner/analysis/parser/BluePearlParser.java
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/BluePearlParser.java:36]: I think adding an own method that set the severity based on the character makes more sense.

**Comment by SimonMatthews-BP** [comment]
Finally, it passes all the tests and the severities are being set in a deterministic and meaningful manner. 

I would appreciate a final review and and pull if you think it ready.

**Review by uhafner — APPROVED** [review]
Thank you very much!

**Comment by SimonMatthews-BP** [comment]
I started with a fresh clone of warnings-ng-plugin-devenv (I puled down my branch, but before doing so, I had rebased it against the origin). I built it as part of clone-repos.sh, started jenkins with "./jenkins.sh". 

I see the BluePearl files in the download, but I still don't see my parser in the list. Can you help me figure out what is missing?

**Comment by KalleOlaviNiemitalo** [comment]
Which list? Do you mean the snippet generator?

**Comment by SimonMatthews-BP** [comment]
No, I mean the list of "Static Analysis Tools", when I select "Record Compiler warnings and static analysis results" as a post build action.

**Comment by SimonMatthews-BP** [comment]
I have an idea what might be missing, but I am out at the moment, so I
can't test it

Simon

**Comment by uhafner** [comment]
Yes, may plan was to remove the `SUPPORTED-FORMATS.md` from the warnings plugin and move it here. But it makes sense to have both listings active, since the warnings plugin file contains the step names as well...

**Pull request merged**
